### ISSUE TITLE
Ability to swipe between days/weeks/months (depending on view)

### ISSUE THREAD

**Issue by insberr** [schedule-x/schedule-x#629]
I am making a scheduling website, which will primarily be used on mobile. One thing I notice and quite enjoy with calendar apps like Google calendar, is being able to swipe left or right to switch between days/weeks/months depending on the view. It would greatly improve the usage for mobile users, especially on small screens where the buttons that do this get hidden.

**Additional context**  
One thing I noticed that could be problematic is when the draggable events plugin is used and when scrolling on mobile, if you start scrolling - finger first touches the screen - on an event, it'll move the event rather than scrolling (TODO: publish a bug report for this?)

**Things that could speed up the process**
- [ ] I can picture implementing at least parts of this feature
- [ ] I believe I know how to do this, and am willing to implement the whole thing.

**Comment by tomosterlund** [comment]
> if you start scrolling - finger first touches the screen - on an event, it'll move the event rather than scrolling (TODO: publish a bug report for this?)

This is behavior I'm aware of, but I'm not quite sure how to go about having D&D enabled, and let someone scroll while pressing an event. I tried to do some workarounds for this already, but found none that made sense.

For now I think the best way is to either have the users get used to the fact that they need to press somewhere outside of an event to scroll, or disable D&D if on mobile

**Comment by insberr** [comment]
I think one way might be a tap and hold kind of solution? Of course that would need to be mobile only, since desktop scrolling works differently. And last I remember doing stuff like that on websites is kinda tough.

**Comment by insberr** [comment]
I found and modified a code snippet from [here](https://codepen.io/deepakkadarivel/pen/LrGEdL). It works well, but I am unsure if it could be implemented into the existing D&D plugin, especially because it only works on touchscreen devices.

Here's my modified code. The element must be held for 200ms before it will become moveable. There is also an `e.preventDefault()` in `moveFn` which prevents touch-scrolling while moving the element. Touch-scrolling still functions normally if you start the touch on the element but do not remain put for the 200ms.
`moveFn` would also need some math magic so the element doesn't jump when it starts being moved, but that's pretty simple (maybe ? I'm no math genius haha)
```js
const HoldDuration = 200; // milliseconds

window.onload = () => {
  var box = document.getElementById('box');

  const moveFn = (e) => {
    // Prevent scrolling while user is moving the element
    e.preventDefault();

    // grab the location of touch
    var touchLocation = e.targetTouches[0];
  
    // assign box new coordinates based on the touch.
    box.style.left = touchLocation.pageX + 'px';
    box.style.top = touchLocation.pageY + 'px';
  };
  
  // Add event listener fot touchstart
  box.addEventListener('touchstart', (e_touchstart) => {
    let timeout;
    const rmTime = () => {
      clearTimeout(timeout)
    };

    // If the user starts dragging before the timeout, clear the timeout.
    box.addEventListener('touchmove', rmTime);

    timeout = setTimeout(() => {
      // Timeout done, element held long enough, dragging is allowed now, remove the event listener that clears the timeout
      box.removeEventListener('touchmove', rmTime);
      
      // Handle movement
      box.addEventListener('touchmove', moveFn);

      // Handle done
      box.addEventListener('touchend', (e_touchend) => {
        // Remove event listener so the cycle can start again
        box.removeEventListener('touchmove', moveFn);
      });
    }, HoldDuration);
  });
}
```

**Comment by nstringham** [comment]
I would be fine if swiping between months and drag and drop where mutually exclusive. I have a read only calendar predominantly used on mobile.

**Comment by insberr** [comment]
> I would be fine if swiping between months and drag and drop where mutually exclusive. I have a read only calendar predominantly used on mobile.

I agree! I think that was my original thought: a new plugin for swiping between day/week/month. I was just mentioning something I noticed that would be conflicting between said new plugin and the existing D&D one if used together.

**Comment by klao** [comment]
We also need something like this for our project.

It's not difficult to create something that works. I have a PoC here: https://github.com/klao/schedule-x/commit/efffff2
You can try it live here: https://mihaly.barasz.com/angcal/ (swiping only works in the day and week view).

It's not straightforward to me how to turn this idea into a plugin, I haven't looked at the plugins yet, but I can give it a try. It's just not clear how "integrated" are plugins… (Would modifying core code to make hooking this up easier be ok, or should I aim for the plugin to be completely independent? How much may I rely on the internal structure—like DOM elements and classes—to stay stable?) Things like this. @tomosterlund do you have a view on it?

**Comment by tomosterlund** [comment]
very cool! Looks like a good approach I think with tracking the diff in pixels and then updating once that diff is high enough. Actually if we can keep the total implementation < 100 lines of code for all mobile compatible views (week, day, month-agenda), I don't mind if we add it directly to `@schedule-x/calendar`.

It is really a part of the basics for how to navigate on mobile, so I guess it would be nice to have directly in the core of the calendar. Maybe we can even just have one implementation for all views, and do it directly on `.sx__view-container`?

Would you mind having a shot at it?

**Comment by HunterLiddell** [comment]
Hi guys, how far along is the development of this task? I have a use-case where I would love to have the user click and hold for n-ms (mobile) before being able to move an event. 

Any update would be great!

**Comment by tomosterlund** [comment]
to my knowledge no one is working on this currently
